webgeocalc/vars.py

```diff
@@ -229,3 +217,5 @@
     'LOCMAX',
     'LOCMIN',
 ]
+
+TYPES = globals()
```

Comment:
Actually… (again 😅) `@SetterProptery` could take an optional `valid_params` directly!
This will simplify both implementation.

There is a new proposal to merge `SetterProperty` and `check_params`:
```python
def set_params(func=None, *, valid_params=None):
    '''Parameter checker decorator.'''

    def decorator(func):
        '''Decolator checker.'''

        @wraps(func)
        def wrapper(_self, value):
            if value not in VALID_PARAMETERS[valid_params]:
                raise CalculationInvalidAttr(
                    name=valid_params,
                    attr=value,
                    valids=VALID_PARAMETERS[valid_params],
                )
            return func(_self, value)
    
        return wrapper

    return decorator(func) if callable(func) else decorator

...

@set_params(valid_params='CALCULATION_TYPE')
def calculation_type(self, val):
    ...

@set_params
def kernels(self, kernel_sets):
    ...
```

I really like this new approach! It's much cleaner 😃

Reply:
Up to just before this point I followed you on how it would work, but for that last one, this one step too far with my current knowledge of decorators 😅

Reply:
but I agree, that it looks much cleaner 👍

Reply:
I added a small description in the proposal above (and fixed the issue when the `valid_params` is missing).
Basically it add an optional argument to the decorator (which is exactly what we need).

Reply:
yes, but does this really replace the `@SetterParam` code?

```
class SetterProperty:
    '''Setter property decorator.'''

    def __init__(self, func, doc=None):
        self.func = func
        self.__doc__ = doc if doc is not None else func.__doc__

    def __set__(self, obj, value):
        return self.func(obj, value)
```

Reply:
Fix: I renammd the `wrapper` to `fset` and now the `decorator` will return a `property`.

Reply:
I'm happy if this works, this is too far, too high for me 😅

Reply:
See [here](https://realpython.com/primer-on-python-decorators/#more-real-world-examples) if you want more  information about optional argument in a decorator.

Reply:
Implement in 39765b4